format/leveldb/leveldb_table.go

```diff
@@ -0,0 +1,443 @@
+				if valueCallbackFn == nil {
+					d.FieldUTF8("value", int(valueLength))
+				} else {
+					d.FieldStruct("value", valueCallbackFn)
+				}
+			})
+		}
+	})
+}
+
+func readBytesWithoutChangingPosition(nBytes int64, d *decode.D) []byte {
+	var result []byte
+	d.RangeFn(d.Pos(), nBytes*8, func(d *decode.D) {
+		br := d.RawLen(d.BitsLeft())
+		result = d.ReadAllBits(br)
+	})
+	return result
+}
+
+// https://github.com/google/leveldb/blob/main/table/format.cc#L24
+func readBlockHandle(d *decode.D) blockHandle {
+	return blockHandle{
+		offset: d.FieldULEB128("offset"),
+		size:   d.FieldULEB128("size"),
+	}
+}
+
+// Read bytes and prefix with given bytes;
+// name read bytes "suffix" and the merged bytes "full" (synthetic field).
+func readPrefixedBytes(prefixBytes []byte, nBytes int, d *decode.D) error {
+	br, err := d.TryFieldRawLen("suffix", int64(nBytes)*8)
+	if err != nil {
+		return err
+	}
+	full := append(prefixBytes, d.ReadAllBits(br)...)
+	d.FieldValueStr("full", string(full), strInferred)
+	return nil
+}
+
+// An "internal key" consists of the triple (user_key, type, sequence_number).
+// https://github.com/google/leveldb/blob/main/db/dbformat.h#L171
+func readInternalKey(sharedBytes []byte, unsharedSize int, d *decode.D) error {
+	// In the LevelDB encoding, the internal key can be cut at any byte:
+	// including the user_key, type, or sequence_number:
+	// https://github.com/google/leveldb/blob/main/table/block_builder.cc#L79-L83
+	//
+	// The resulting prefix is then shared among subsequent keys and not
+	// specified explicitly by them. Here, we handle each cutoff case.
+	//
+	// All sizes are in bytes unless mentioned otherwise.
+	keySize := len(sharedBytes) + unsharedSize
+	typeAndSequenceNumberSize := 8
+	//                        key
+	// +-----------------------------------------------+
+	//               user_key
+	// +---------------------------------+
+	//               ⁞   user_key_suffix   type sequence_number
+	// [AAAAAAAAAAAA]⁞[BBBBBBBBBBBBBBBBBB] [T] [SSSSSSS]
+	//               ⁞                      1   7 bytes
+	// +------------+⁞+--------------------------------+
+	//     shared    ⁞             unshared
+	//               ⁞
+	//             cutoff
+	if keySize < typeAndSequenceNumberSize || int64(unsharedSize) > d.BitsLeft()/8 {
+		return fmt.Errorf("key size %d or unshared size %d invalid", keySize, unsharedSize)
+	}
+
+	// case 1: user_key, type, and sequence_number fit fully in unshared.
+	if len(sharedBytes) == 0 {
+		d.FieldUTF8("user_key", keySize-typeAndSequenceNumberSize)
+		d.FieldU8("type", valueTypes, scalar.UintHex)
+		d.FieldU56("sequence_number")
+		return nil
+	}
+
+	// case 2: type and sequence_number fit fully in unshared: simulate user_key value.
+	if unsharedSize >= typeAndSequenceNumberSize {
+		br := d.FieldRawLen("user_key_suffix", int64(unsharedSize-typeAndSequenceNumberSize)*8)
+		d.FieldValueStr("user_key", string(append(sharedBytes, d.ReadAllBits(br)...)), strInferred)
```

Comment:
Could it be something like?
```go
suffix := d.FieldUTF8("user_key_suffix", int64(unsharedSize-typeAndSequenceNumberSize))
d.FieldValueStr("user_key", string(sharedBytes)+suffix), strInferred)
```

An alternativ to using `d.FieldValueStr` could be to only have one `user_key` field and let the symbolic value for that one string be the inferred string. But i think i like how it is now which i guess is more true to how thing actually work, that it's a suffix, then and just add a bit extra to make it convenient.

btw `append(sharedBytes, d.ReadAllBits(br)...)` can something in go be a bit tricky, if i remember correctly the `sharedBytes` slice might end up being reused if it as capacity, but in this case it should be fine i think as it get turned into a new string? (which immutable and not share the slice bytes)

Reply:
@wader I like your solution.
"the sharedBytes slice might end up being reused"
the `append()` of Go I don't quite understand yet. Apparently it is an amoritized-O(1) operation yet it seems sharedBytes doesn't get mutated? (So it's not doing `.append(....)` like in Python or `.push` like in JavaScript.)

Reply:
@wader Although, on second thought, if the user_key is cut at some UTF-8 multi-byte character, this might fail. However, maybe it is good enough.

Reply:
In testing, I see it does indeed fail:

![image](https://github.com/wader/fq/assets/216956/ec9c296d-68d0-430c-a324-1993a8d18587)

Reply:
In LevelDB, both keys and values are bytestrings; so a UTF-8 encoding would typically be used, but most not necessarily.

Reply:
@mikez I made some own functions for now to handle this. See recent commit.